### PR TITLE
S3C-287 FIX: Always set bucket location

### PR DESCRIPTION
**Pull request by nicolas2bert**
We should always explicitly set a location constraint on a bucket if data backend equals 'multiple'.

### CODE REVIEW

File: lib/api/bucketPut.js [hunk 1]
```diff
@@ -8,6 +8,52 @@ import config from '../Config';
 import aclUtils from '../utilities/aclUtils';
 import { pushMetric } from '../utapi/utilities';
 
+const { locationConstraints, restEndpoints, backends } = config;
+
+
+/**
+ * _checkLocationConstraint - check that a location constraint is explicitly
```
Comment (LaurenSpiegel) [inline lib/api/bucketPut.js]: no underscore

File: lib/api/bucketPut.js [hunk 2]
```diff
@@ -30,12 +76,24 @@ function _parseXML(request, log, cb) {
...
+            if (locationCheck.error) {
+                return cb(locationCheck.error);
+            }
+            return cb(null, locationCheck.locationConstraint);
         });
     }
-    // We need the second parameter to be `undefined` so the waterfall maintains
-    // the position of the `next` argument.
-    return process.nextTick(() => cb(null, undefined));
+    // We need the second parameter to be empty string so the waterfall
```
Comment (LaurenSpiegel) [inline lib/api/bucketPut.js]: I don't think you need this comment anymore. We are sending either the default or an empty string now.

File: lib/api/bucketPut.js [hunk 3]
```diff
@@ -8,6 +8,52 @@ import config from '../Config';
...
+ * @return {undefined}
+ */
+
+export function checkLocationConstraint(request, locationConstraint, log) {
+    // AWS JS SDK sends a request with locationConstraint us-east-1 if
+    // no locationConstraint provided.
+    const { parsedHost } = request;
+    let locationConstraintChecked;
+    if (locationConstraint) {
+        locationConstraintChecked = locationConstraint;
```
Comment (LaurenSpiegel) [inline lib/api/bucketPut.js]: what happened to check to make sure the locationConstraint is one listed in the config?

File: lib/api/bucketPut.js [hunk 4]
```diff
@@ -8,6 +8,52 @@ import config from '../Config';
...
+        locationConstraintChecked = restEndpoints[parsedHost];
+    } else if (backends.data === 'multiple') {
+        const errMsg = 'no location constraint explicitly set on a' +
+          ' bucket';
+        log.trace(`locationConstraint is invalid - ${errMsg}`);
+        return { error: errors.InvalidLocationConstraint.
+          customizeDescription(errMsg) };
+    } else {
+        // return empty string location if no location constraint explicitly set
+        // and data backend not set to "multiple"
```
Comment (LaurenSpiegel) [inline lib/api/bucketPut.js]: add note: THIS IS ONLY FOR THE OPEN SOURCE VERSION THAT IS RUNNING IN FILE OR MEM.

File: lib/api/bucketPut.js [hunk 5]
```diff
@@ -8,6 +8,52 @@ import config from '../Config';
...
 
+const { locationConstraints, restEndpoints, backends } = config;
+
+
+/**
+ * _checkLocationConstraint - check that a location constraint is explicitly
+ * set on the bucket and the value of the location is not listed in the
+ * locationConstraint config.
+ * Note: if data backend equals "multiple", you must set a location constraint
+ * @param {string} locationConstraint - location constraint
```
Comment (nicolas2bert) [inline lib/api/bucketPut.js]: yes - thanks

File: lib/api/bucketPut.js (hunk 4 above)
Comment (nicolas2bert) [inline lib/api/bucketPut.js]: done :)

**Comment by LaurenSpiegel** [comment]
@ironman-machine go

**Pull request merged**